source/extensions/filters/http/cache/config.cc

```diff
@@ -6,27 +6,39 @@ namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace Cache {
-
-Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
-    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
-    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+namespace {
+std::shared_ptr<HttpCache>
+getCache(const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+         Server::Configuration::CommonFactoryContext& context) {
   const std::string type{TypeUtil::typeUrlToDescriptorFullName(config.typed_config().type_url())};
   HttpCacheFactory* const http_cache_factory =
       Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
   if (http_cache_factory == nullptr) {
     throw EnvoyException(
         fmt::format("Didn't find a registered implementation for type: '{}'", type));
   }
+  return http_cache_factory->getCache(config, context);
+}
+} // namespace
 
-  auto cache = http_cache_factory->getCache(config, context);
-
-  return [config, stats_prefix, &context,
-          cache](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    callbacks.addStreamFilter(std::make_shared<CacheFilter>(config, stats_prefix, context.scope(),
-                                                            context.timeSource(), *cache));
+Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+  auto filter_config = std::make_shared<CacheFilterConfig>(getCache(config, context), config);
```

Comment:
Again, that's not how configs work unfortunately. We *have to* get the cache that's configured at the listener level, because some instance of the filter might not have a per-route cache. (With a future addition of 'disabled' config this can be avoided.)

Also we don't want to be doing getCache during decodeHeaders (which is the first moment at which we know what any possibly-merged config might look like), because that would then involve calling getCache for every connection, which would be expensive. This way, if you have a base config and two route configs, getCache is called only 3 times.

Reply:
A particular bit of config can be handled
1. during CacheFilterFactory's initialization,
2. during CacheFilter's initialization,
3. during decodHeaders (or similar), or
4. some combination of the above.

Envoy policy is that any possible config errors should be handled in step 1, which is why it seems to be standard to convert protos to native C++ data structures in step 1. (CacheFilter doesn't do this, but could.)

As I understand it, your intent is to allow overriding CacheFilter's entire config, which means that both steps 2 and 3 must be able to handle the full config. We can't create the HttpCache implementation until we know which kind to create, so we either need to change how HttpCache creation is done, or we need to step back from supporting the complete config in per-route config.

Reply:
Edit: I'm realizing that there's something I'm not understanding about how per-route config works. I'll take another look at this when I've corrected that.